diff --git a/share/mk/bsd.sysdir.mk b/share/mk/bsd.sysdir.mk
index 52b841a51fea..a35457daafcb 100644
--- a/share/mk/bsd.sysdir.mk
+++ b/share/mk/bsd.sysdir.mk
@@ -15,3 +15,5 @@ SYSDIR=	${_dir:tA}
     !exists(${SYSDIR}/conf/kmod.mk)
 .error Unable to locate the kernel source tree. Set SYSDIR to override.
 .endif
+
+ZFSTOP=${SYSDIR}/contrib/subrepo-openzfs
diff --git a/sys/vm/vm_mmap.c b/sys/vm/vm_mmap.c
index a3742587c0a2..224ae325d991 100644
--- a/sys/vm/vm_mmap.c
+++ b/sys/vm/vm_mmap.c
@@ -1890,6 +1890,9 @@ vm_mmap_cdev(struct thread *td, vm_size_t objsize, vm_prot_t *protp,
 	if (error != 0)
 		return (error);
 #endif
+	*protp = prot = VM_PROT_ADD_CAP(*protp);
+	*maxprotp = VM_PROT_ADD_CAP(*maxprotp);
+
 	/*
 	 * First, try d_mmap_single().  If that is not implemented
 	 * (returns ENODEV), fall back to using the device pager.
@@ -1900,8 +1903,14 @@ vm_mmap_cdev(struct thread *td, vm_size_t objsize, vm_prot_t *protp,
 	 * XXX assumes VM_PROT_* == PROT_*
 	 */
 	error = dsw->d_mmap_single(cdev, foff, objsize, objp, (int)prot);
-	if (error != ENODEV)
+	if (error != ENODEV) {
+		if (error == 0) {
+			VM_OBJECT_WLOCK(*objp);
+			vm_object_set_flag(*objp, OBJ_HASCAP);
+			VM_OBJECT_WUNLOCK(*objp);
+		}
 		return (error);
+	}
 	obj = vm_pager_allocate(OBJT_DEVICE, cdev, objsize, prot, *foff,
 	    td->td_ucred);
 	if (obj == NULL)
